Service Level Agreement

sirkrypt0 opened the pull request:
TODO:

- [x] Maybe some more fixed test cases to kick out some wrong solutions
- [ ] Implement solution by Aneja et al. and add it to the solution notes
- [x] Check time bounds

Closes #15

sirkrypt0 committed:
sla: add problem with basic solution and test generation

roehrt commented on cp23/hpi-fun2/sla/executables/generator.py:
```diff
@@ -0,0 +1,71 @@
+TESTCASES = [
+    ('sample1', SAMPLE1, 'Sample 1'),
+    ('sample2', SAMPLE2, 'Sample 2'),
+    ('large', generate_random_test_case(50, 1000, R_MAX), 'Sample filling n,m bounds'),
```
```suggestion
    ('large', generate_random_test_case(N_MAX, M_MAX, R_MAX), 'Sample filling n,m bounds'),
```

roehrt commented on cp23/hpi-fun2/sla/executables/generator.py:
```diff
@@ -0,0 +1,71 @@
+TESTCASES = [
+    ('sample1', SAMPLE1, 'Sample 1'),
+    ('sample2', SAMPLE2, 'Sample 2'),
+    ('large', generate_random_test_case(50, 1000, R_MAX), 'Sample filling n,m bounds'),
```
```suggestion
    ('large', generate_random_test_case(N_MAX, M_MAX, R_MAX), f'Maximum input n = {N_MAX}'),
```
Is it meant to be a sample?

sirkrypt0 replied:
Since the name doesn't start with `sample`, I assume it's not a sample? At least looking at the output in the `build/judge-package/data/sample` folder it seems that the large test case is not part of the samples but rather the secret test cases.

roehrt replied:
Sure, but your testcase sample description said "sample"

roehrt commented on cp23/hpi-fun2/sla/problem.tex:
```diff
@@ -0,0 +1,32 @@
+
+The first line contains $n, m ~ (2 \leq n \leq 50; 0 \leq m \leq 1000)$. The next $m$ lines
+contain integers $a, b, r$ denoting a bidirectional link between $a$ and $b$ with data rate $r ~ (1 \leq a, b \leq n;
+1 \leq r \leq 10^6)$ in $MiB / s$.
```
Personally I wouldn't typeset unit in math variable font but that's up to you.

roehrt commented on cp23/hpi-fun2/sla/executables/generator.py:
```diff
@@ -0,0 +1,71 @@
+for i in range(1, 21):
+    print(f'Generating random case {i}')
+    Path(f'random{i}.in').write_text(generate_random_test_case())
+    Path(f'random{i}.desc').write_text(f'random testcase {i}\n')
```
[Optional] Add the n and m to the testcase description.

roehrt reviewed (commented):
Very clean generator, solution and problem statement :raised_hands:
Some things left to do:

- [x] As you already listed for yourself: some more small edges cases would be nice
- [x] Add an Edmonds-Karp solution for me for testing the timelimit on the :judge: since Push-Relabel wasn't part of CP1 (sadly)

roehrt commented on cp23/hpi-fun2/sla/executables/generator.py:
```diff
@@ -0,0 +1,71 @@
+def generate_random_test_case(n = None, m = None, r = None):
+    n = n if n is not None else random.randint(N_MIN, N_MAX)
+    max_possible_edges = n * (n-1) / 2
+    m = m if m is not None else random.randint(M_MIN, min(max_possible_edges, M_MAX))
```
```suggestion
    m = min(max_possible_edges, m if m is not None else random.randint(M_MIN, M_MAX))
```

roehrt committed:
Add solution tobi

roehrt committed:
Allow no edges

sirkrypt0 committed:
sla: fix minimum m calculation in generator and tweak sample description

sirkrypt0 committed:
sla: don't typeset unit in problem description

sirkrypt0 committed:
sla: add m=0 and no path test cases and adjust problem accordingly

We adjust the problem statement to better reflect (apart from the
input definition with 0 <= m ...) that the two datacenters may not
be connected at all.

sirkrypt0 committed:
sla: add edmonds-karp solution

sirkrypt0 commented:
I added an Edmonds-Karp solution now :)

roehrt commented on cp23/hpi-fun2/sla/executables/generator.py:
```diff
@@ -0,0 +1,89 @@
+    print(f'Generating random case {i}')
+
+    n = random.randint(N_MIN, N_MAX)
+    # While the result is theoretically always an integer, Python still represents it as a float.
+    # This results in the m below possibly becoming a float, which the `range(m)` in
+    # generate_random_test_case doesn't like.
+    # Hence, we convert to an int explicitly without loosing precision.
+    max_possible_edges = int(n * (n-1) / 2)
```
```suggestion
    max_possible_edges = n * (n-1) // 2
```

roehrt commented on cp23/hpi-fun2/sla/executables/generator.py:
```diff
@@ -0,0 +1,89 @@
+M_MAX = 1000
+
+def generate_random_test_case(n, m, r = None):
+    max_possible_edges = n * (n-1) / 2
```
```suggestion
    max_possible_edges = n * (n-1) // 2
```

roehrt reviewed (approved):
![sla](https://github.com/compprog-lecture-tools/fun-contest-2023/assets/33495148/ccce7502-7ee7-4128-ba3a-4201b093d467)
It runs on the :judge:! :tada:

roehrt committed:
Be a cool kid

sirkrypt0 committed:
sla: use integer division instead of conversion

roehrt commented:
If you're done now, this looks mergable :tada: So let me if you are still working on it otherwise I'd press the green button :)

sirkrypt0 commented:
Awesome! I'll let you know. I'm still working on that solution from the paper :)

sirkrypt0 committed:
sla: add solution implementing the paper from Aneja et al

sirkrypt0 commented:
I finally got it (after I figured out I should use float for the lambda value ...) and added the solution from their paper. I'd be interested if it actually runs faster on the judge and by what factor. At least my tests show that the amount of max flow calculations is reduced (in some cases), so I'd expect it to run faster.

Maybe I did something wrong, but I can't verify their claim "Lemma 3: The algorithm Newton stops with an optimal solution in at most |N| iterations.". In testcase random 4 I record 34 iterations, whereas N is 7.

Also, I have this weird edge case when `q(lambda)` becomes 0 (happens in random17) I haven't investigated further yet.

roehrt commented on cp23/hpi-fun2/sla/executables/solution.aneja.cpp:
```diff
@@ -0,0 +1,159 @@
+  vector<int> S, S_lambda;
+  int v = findMinCut(links, n, INT_MAX, S, q);
+
+  float lambda = INT_MIN;  // Initialize lambda with a very large value
+  // Find the initial lambda value
+  for (auto& l : S) {
+    lambda = max(lambda, float(links[l].r));
+  }
+
+  int v_lambda = findMinCut(links, n, lambda, S_lambda, q_lambda);
+
+  // Iterate until v(lambda) >= max_flow_value
+  while (v_lambda < max_flow_value && q_lambda != 0) {
+    lambda = lambda + (max_flow_value - v_lambda) / float(q_lambda);
```
I think this can be rewritten using only integers. Since `lambda` will be casted to `int` anyways, you could keep track of the denominator and numerator. This would be faster and more precise.

roehrt commented on cp23/hpi-fun2/sla/executables/solution.aneja.cpp:
```diff
@@ -0,0 +1,159 @@
+  vector<int> S, S_lambda;
+  int v = findMinCut(links, n, INT_MAX, S, q);
+
+  float lambda = INT_MIN;  // Initialize lambda with a very large value
```
```suggestion
  float lambda = -numeric_limits<float>::infinity();
```

roehrt reviewed (commented):
I'll revisit the source code with the paper in mind. In the meantime that is what the :judge: says:
![Screenshot from 2023-07-29 15-39-56](https://github.com/compprog-lecture-tools/fun-contest-2023/assets/33495148/5de6aaf2-3f8e-4ce6-b51a-819ab08d8f67)
(Timelimit on "Sample with edges but no s-t-path")

roehrt commented:
Hmm.. Maybe I misunderstood your solution but your referenced paper presents two different problems. We are here considering the "most vital arc" problem, right? If I skimmed the paper correctly, we can limit our search to edges that are part of any maximum (paper says maximal but I assume they mean maximum) spanning tree.

sirkrypt0 commented:
> Hmm.. Maybe I misunderstood your solution but your referenced paper presents two different problems. We are here considering the "most vital arc" problem, right? If I skimmed the paper correctly, we can limit our search to edges that are part of any maximum (paper says maximal but I assume they mean maximum) spanning tree.

Yeah, you might be right. I actually implemented the maximum residual flow solution... Your comment made me revisit the introduction and I realised the difference in the problems again. Especially, that the maximum residual flow need not be integral (although they have an extension which covers that) convinces me that my problem really asks more for the most vital arc.

roehrt commented:
You can still implement that approach if you wish to. We could also merge it any time so no time pressure here :)
